proxy/upstream.go

```diff
@@ -0,0 +1,114 @@
+package proxy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/coinbase/redisbetween/config"
+	"github.com/coinbase/redisbetween/redis"
+	"github.com/coinbase/redisbetween/utils"
+	"go.uber.org/zap"
+	"sync"
+)
+
+type UpstreamManager interface {
+	Add(ctx context.Context, c *config.Upstream) error
+	ConfigByName(ctx context.Context, name string) (*config.Upstream, bool)
+	LookupByName(ctx context.Context, name string) (redis.ClientInterface, bool)
+	Shutdown(ctx context.Context) error
+}
+
+func NewUpstreamManager() UpstreamManager {
+	return &upstreamManager{
+		configsByName:   make(map[string]*config.Upstream),
+		uniqueUpstreams: make(map[string]redis.ClientInterface),
+		upstreamsByName: make(map[string]redis.ClientInterface),
+	}
+}
+
+type upstreamManager struct {
+	configsByName   map[string]*config.Upstream
+	uniqueUpstreams map[string]redis.ClientInterface
+	upstreamsByName map[string]redis.ClientInterface
+	sync.RWMutex
+}
+
+func (m *upstreamManager) Add(ctx context.Context, u *config.Upstream) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// Make a local copy of the config
+	cfg := &config.Upstream{}
+	*cfg = *u
```

Comment:
Just pass `u` by value to this function instead of by reference

Reply:
done